### PR TITLE
Fix specifying gpu_id, add tests.

### PR DESCRIPTION
**Pull request by trivialfis**
Address #3850 .

I don't have the multi-gpu machine for testing currently.

### CODE REVIEW

**Comment by hcho3** [comment]
@trivialfis We may want to list this as a known issue. Do you have a short description of the issue?

**Comment by hcho3** [comment]
@trivialfis Got it. I'll trust your judgment on this.

**Comment by trivialfis** [comment]
@hcho3 Tested with the R script provided by @joegaotao . Thanks.  I think it should now be fixed.

@RAMitchell But it's not clear to me, what's the point of having `normalised_device_index`.  Can we just use the device index obtained from CUDA?

**Comment by trivialfis** [comment]
@hcho3 Not really.  Just talked to @RAMitchell , the current fix will disable dividing training sessions across GPUs.   Although the script won't fail but it defeats the purpose of specifying `gpu_id`.

**Comment by trivialfis** [comment]
Let me keep trying.

**Comment by trivialfis** [comment]
@RAMitchell Ready for a review.  I haven't figure out how to make a unified way to allocate GPU, but the current fix won't add any burden to the code base, see if we can merge it.

File: src/common/common.h [hunk 1]
```diff
@@ -169,14 +169,15 @@ class GPUSet {
...
-  }
+
   /*! \brief Counting from gpu_id */
   GPUSet Normalised(int gpu_id) const {
+    int n_devices_visible = AllVisible().Size();
+    CHECK(Size() + gpu_id <= n_devices_visible)
+        << "n_gpu + gpu_id should less than or equal to total number of available devices."
+        << "\n n_gpu: " << Size()
+        << ", gpu_id: " << gpu_id
+        << ", number of available devices: " << n_devices_visible;
```
Comment (RAMitchell) [inline src/common/common.h]: Lets remove Unnormalised(), I think it is wrong as the range can be larger than the number of GPUs. This unnormalised/normalised terminology is confusing and I can see it somehow has different meanings across different parts of the code. From now on whenever a device index is referred to it means the physical device ordinal of the GPU.

File: src/common/common.h [hunk 2]
```diff
@@ -169,14 +169,15 @@ class GPUSet {
...
-  /*! \brief Ensure gpu_id is correct, so not dependent upon user knowing details */
-  static int GetDeviceIdx(int gpu_id) {
-    auto devices = AllVisible();
-    CHECK(!devices.IsEmpty()) << "Empty device.";
-    return (std::abs(gpu_id) + 0) % devices.Size();
-  }
+
   /*! \brief Counting from gpu_id */
   GPUSet Normalised(int gpu_id) const {
+    int n_devices_visible = AllVisible().Size();
```
Comment (RAMitchell) [inline src/common/common.h]: Let's rename this function , without unnormalised/normalised terminology.

File: src/common/common.h [hunk 3]
```diff
@@ -197,7 +198,9 @@ class GPUSet {
   bool IsEmpty() const { return Size() == 0; }
   /*! \brief Get un-normalised index. */
```
Comment (RAMitchell) [inline src/common/common.h]: As above let's remove the function Index().

File: src/tree/updater_gpu_hist.cu [hunk 4]
```diff
@@ -251,15 +251,15 @@ struct DeviceHistogram {
   thrust::device_vector<GradientPairSumT::ValueT> data;
   const size_t kStopGrowingSize = 1 << 26;  // Do not grow beyond this size
   int n_bins;
-  int device_idx;
+  int normalised_device_idx;
```
Comment (RAMitchell) [inline src/tree/updater_gpu_hist.cu]: This should all be device_idx, as before let's remove all normalised/unnormalised terminology.

File: src/common/transform.h [hunk 5]
```diff
@@ -125,19 +125,22 @@ class Transform {
...
+
+      // Extract index to deal with possible old OpenMP.
+      size_t device_beg = *(devices.begin());
+      size_t device_end = *(devices.end());
 #pragma omp parallel for schedule(static, 1) if (devices.Size() > 1)
-      for (omp_ulong i = 0; i < devices.Size(); ++i) {
-        int d = devices.Index(i);
+      for (omp_ulong device = device_beg; device < device_end; ++device) {  // NOLINT
         // Ignore other attributes of GPUDistribution for spliting index.
+        int d_unnormalised = devices.Index(device);
```
Comment (RAMitchell) [inline src/common/transform.h]: In situations like this you need to have some index beginning from 0 you could get it from the enclosing for loop:
```
 for (omp_ulong i = 0; i < devices.Size(); ++i) {
    auto device_idx = devices[i];
    size_t shard_size =
            GPUDistribution::Block(devices).ShardSize(range_size, i);
```

File: src/tree/updater_gpu_hist.cu [hunk 6]
```diff
@@ -1336,6 +1338,7 @@ class GPUHistMaker : public TreeUpdater {
   common::Monitor monitor_;
   dh::AllReducer reducer_;
   std::vector<ValueConstraint> node_value_constraints_;
+  /*! List storing normalised device index*/
```
Comment (RAMitchell) [inline src/tree/updater_gpu_hist.cu]: We should be able to remove this.

**Comment by trivialfis** [comment]
@hcho3 I can't address these problems today.  I need some more time to completely remove the normalise/unnormalise stuffs.  So known issue it is.

**Comment by hcho3** [comment]
@trivialfis Okay, I'll add it to the known issues. How should I summarize this issue (in a sentence)? How about "`gpu_id` cannot be set when one GPU is used"?

**Comment by trivialfis** [comment]
"Specifying gpu-id is not yet supported." is more realistic from my perspective.

**Comment by trivialfis** [comment]
Due to  #3794 , even if we try to support specifying `gpu_id`, users still need to have a copy their dataset for each GPU.

**Comment by trivialfis** [comment]
@RAMitchell Ready for another review now. :)

File: src/common/common.h [hunk 7]
```diff
@@ -147,61 +147,85 @@ struct AllVisibleImpl {
...
+  /*
+   * By default, we have two configurations of identifying device, one
+   * is the device id obtained from `cudaGetDevice'.  But we sometimes
+   * store objects that allocated one for each device in a list, which
+   * requires a zero-based index.
+   *
+   * Hence, `DeviceId' converts a zero-based index to actual device id,
+   * `Index' converts a device id to a zero-based index.
+   */
+  GpuIndex DeviceId(GpuIndex index) const {
```
Comment (RAMitchell) [inline src/common/common.h]: Suggest index is NOT of type GpuIndex. I assume GpuIndex is intended to refer to a physical device?

File: src/common/common.h [hunk 8]
```diff
@@ -147,61 +147,85 @@ struct AllVisibleImpl {
...
+    GpuIndex result = *devices_.begin() + index;
+    CHECK(Contains(result)) << "\nDevice " << result << " is not in GPUSet."
+                            << "\nIndex: " << index
+                            << "\nGPUSet: (" << *begin() << ", " << *end() << ")"
+                            << std::endl;
+    return result;
+  }
+  GpuIndex Index(GpuIndex device) const {
```
Comment (RAMitchell) [inline src/common/common.h]: As above maybe this function should return an integer and not GpuIndex? You should be clear about what this type is and not use it for both these purposes.

File: src/common/common.h (hunk 7 above)
Comment (trivialfis) [reply src/common/common.h]: Done. :)

File: src/common/common.h (hunk 8 above)
Comment (trivialfis) [reply src/common/common.h]: Renamed `GpuIndex` to `GpuIdType', and use it along with `size_t` to indicate different types of indices.

**Comment by hcho3** [comment]
@trivialfis I just added a new multi-GPU worker, but somehow I broke the old one. Let me reboot it.

**Comment by trivialfis** [comment]
@hcho3 Thanks. :)

**Comment by hcho3** [comment]
@trivialfis Both workers are now up and running.

**Pull request merged**